File: Array/T496_Next Greater Element I/GO/main.go
package main

import "fmt"

func main() {
	nums1 := []int{4, 1, 2}
	nums2 := []int{1, 3, 4, 2}
	fmt.Println("case1:", nextGreaterElement(nums1, nums2))

}

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	table := map[int]int{}
	stack := []int{}
	nums2Len := len(nums2)
	for i := nums2Len - 1; i >= 0; i-- {
		// 先出栈
		for len(stack) > 0 && stack[len(stack)-1] <= nums2[i] {
			stack = stack[:len(stack)-1]
		}
		//记录到哈希表
		if len(stack) > 0 {
			table[nums2[i]] = stack[len(stack)-1]
		} else {
			table[nums2[i]] = -1
		}
		stack = append(stack, nums2[i])
	}

	res := []int{}
	for _, v := range nums1 {
		res = append(res, table[v])
	}
	return res
}

File: Array/T496_Next Greater Element I/PYTHON/main.py
from typing import List
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # 单调栈 + 哈希表,用空间换取时间，nums2中，从后向前遍历，计算出每一个值x，第一个比x大的值y,存储到哈希表中
        h_table = {}
        # 栈
        stack = []
        for value in reversed(nums2):
            # 将比当前value小的元素出栈
            while stack and stack[-1]<=value:
                stack.pop()
            h_table[value] = stack[-1] if stack else -1
            # 将当前元素入栈
            stack.append(value)
        # 查询哈希表获取结果
        return [h_table[value] for value in nums1]